### PR TITLE
Cache and query improvements for BannersModelBanner::getItem()

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes

Improve the cache use in `BannersModelBanner::getItem()` and remove a duplicated query.

1) Use the callback cache controller for reading from the cache instead of the base controller class.
2) Currently this method runs both `$db->execute()` and `$db->loadObject()` resulting in two queries, removes the `$db->execute()` call.
3) Handles cache exceptions and ignores other exceptions.  Right now it's catching all thrown `RuntimeException` objects and calling `JError::raiseError()` which results in the error page being displayed.  So there's really no point in trying to handle other errors here, let those bubble up the stack and hit the global handler which does the same if nothing else catches it (if a caller wanted to they could now have some sort of graceful fallback for other error conditions versus the model directly triggering the error page with no option of doing anything else; models are really a bad place to be calling `JError` at all).

### Testing Instructions

A banner should still be read and used correctly when loaded through this model method.

### CODE REVIEW

File: components/com_banners/models/banner.php [hunk 1]
```diff
@@ -143,43 +143,41 @@ public function &getItem()
...
+							$db->quoteName('a.clickurl', 'clickurl'),
+							$db->quoteName('a.cid', 'cid'),
+							$db->quoteName('a.track_clicks', 'track_clicks'),
+							$db->quoteName('cl.track_clicks', 'client_track_clicks'),
+						)
 					)
-					->from('#__banners as a')
-					->where('a.id = ' . (int) $id)
-
+					->from($db->quoteName('#__banners'), 'a')
```
Comment (joomdonation) [inline components/com_banners/models/banner.php]: Look like a typo here, it should be changed to **->from($db->quoteName('#__banners', 'a'))** . At the moment, when I applied the patch, I received this error. 

> 1054 Unknown column 'a.clickurl' in 'field list'

File: components/com_banners/models/banner.php [hunk 2]
```diff
@@ -143,43 +143,41 @@ public function &getItem()
...
-				}
-
-				$this->_item = $db->loadObject();
-				$cache->store($this->_item, $id);
+				return $db->loadObject();
+			};
+
+			try
+			{
+				$this->_item = $cache->get($loader, array($id), md5(__METHOD__ . $id));
```
Comment (joomdonation) [inline components/com_banners/models/banner.php]: Is it safe to use __METHOD__ . $id. In Joomla at the moment, frontend and backend model has same class name, so I guess if someone use cache to get result from backend BannersModelBanner model, the result might be wrong ? (I haven't used cache in my extensions, just guess it only)

**Comment by joomdonation** [comment]
I have tested this item :red_circle: unsuccessfully on 421dfa629ac8ea28277019f156d3fe256cf27b2c<br /><br />Received error below after applying the patch:

1054 Unknown column 'a.clickurl' in 'field list'
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14138">issues.joomla.org/tracker/joomla-cms/14138</a>.</sub>

**Comment by ladyjer** [comment]
I have tested this item :red_circle: unsuccessfully on 421dfa629ac8ea28277019f156d3fe256cf27b2c<br /><br />Same error here, after click on banner I received:

1054 Unknown column 'a.clickurl' in 'field list'<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14138">issues.joomla.org/tracker/joomla-cms/14138</a>.</sub>

File: components/com_banners/models/banner.php (hunk 2 above)
Comment (mbabker) [inline components/com_banners/models/banner.php]: The admin model doesn't use caching.

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on 71e38e61bc53fb77f395d0b158942f4bc1cfbd04<br /><br />Works OK now, thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14138">issues.joomla.org/tracker/joomla-cms/14138</a>.</sub>

**Comment by ladyjer** [comment]
I have tested this item :white_check_mark: successfully on 71e38e61bc53fb77f395d0b158942f4bc1cfbd04<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14138">issues.joomla.org/tracker/joomla-cms/14138</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC as 2 successfully Tests?

**Comment by dgt41** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14138">issues.joomla.org/tracker/joomla-cms/14138</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
thanks @dgt41

**Pull request merged**